index.php

```diff
@@ -1681,41 +1703,103 @@ public function installUpdateModuleAction(): void
-			$zipFile = $this->filesPath . '/ZIPFromURL.zip';
-			$zipResource = fopen($zipFile, 'w');
-			$ch = curl_init();
-			curl_setopt($ch, CURLOPT_URL, $url);
-			curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
-			curl_setopt($ch, CURLOPT_FILE, $zipResource);
-			curl_exec($ch);
-			$curlError = curl_error($ch);
-			curl_close($ch);
-			$zip = new \ZipArchive;
-			if ($curlError || $zip->open($zipFile) !== true || (stripos($url, '.zip') === false)) {
-				$this->recursiveDelete($this->rootDir . '/data/files/ZIPFromURL.zip');
-				$this->alert('danger',
-					'Error opening ZIP file.' . ($curlError ? ' Error description: ' . $curlError : ''));
+		$zipFile = $this->filesPath . '/ZIPFromURL.zip';
+		$zipResource = fopen($zipFile, 'w');
+		$ch = curl_init();
+		curl_setopt($ch, CURLOPT_URL, $url);
+		curl_setopt($ch, CURLOPT_FOLLOWLOCATION, true);
+		curl_setopt($ch, CURLOPT_FILE, $zipResource);
+		curl_exec($ch);
+		$curlError = curl_error($ch);
+		curl_close($ch);
+		if (is_resource($zipResource)) {
+			fclose($zipResource);
+		}
+
+		$zip = new \ZipArchive;
+		if ($curlError || $zip->open($zipFile) !== true || (stripos($url, '.zip') === false)) {
+			$this->recursiveDelete($zipFile);
+			$this->alert('danger',
+				'Error opening ZIP file.' . ($curlError ? ' Error description: ' . $curlError : ''));
+			$this->redirect();
+		}
+
+		// Basic zip-slip protection: reject any absolute paths or parent traversal.
+		for ($i = 0; $i < $zip->numFiles; $i++) {
+			$name = $zip->getNameIndex($i);
+			if ($name === false) {
+				continue;
+			}
+			if (strpos($name, "\0") !== false
+				|| preg_match('#(^/|^\\\\|^[a-zA-Z]:)#', $name)
+				|| strpos($name, '../') !== false
+				|| strpos($name, '..\\') !== false
+			) {
+				$zip->close();
+				$this->recursiveDelete($zipFile);
+				$this->alert('danger', 'ZIP file contains invalid paths.');
 				$this->redirect();
 			}
-			// First delete old plugin folder
-			$this->recursiveDelete($path . $folderName);
+		}
+
+		// First delete old plugin/theme folder
+		$this->recursiveDelete($path . $folderName);
+
+		// Then extract new one
+		$zip->extractTo($path);
+		$zip->close();
+		$this->recursiveDelete($zipFile);
+
+		$moduleFolder = $path . $folderName . '-master';
+		if (!is_dir($moduleFolder)) {
+			$moduleFolder = $path . $folderName . '-main';
+		}
+		if (is_dir($moduleFolder) && !rename($moduleFolder, $path . $folderName)) {
+			throw new Exception('Theme or plugin not installed. Possible cause: themes or plugins folder is not writable.');
+		}
+
+		// If this plugin belongs to an exclusive group (e.g. editor/translation), disable other enabled plugins in the same group.
+		// Keep disabled state on updates (if the plugin was disabled before updating).
+		if ($type === self::PLUGINS_DIR) {
+			$disabled = $this->getDisabledPlugins();
 
-			// Then extract new one
-			$zip->extractTo($path);
-			$zip->close();
-			$this->recursiveDelete($this->rootDir . '/data/files/ZIPFromURL.zip');
```

Comment:
Since you are creating this zip everytime on install, it would be wise to delete it afterwards?